src/vcpkg/tools.cpp

```diff
@@ -14,21 +16,77 @@
 
 #include <regex>
 
+using namespace vcpkg;
+
+namespace
+{
+
+    DECLARE_AND_REGISTER_MESSAGE(ToolVersionMetadataIllFormed,
+                                 (msg::actual),
+                                 "{actual} is the actual text attempted to be parsed as a version number",
+                                 "Unable to recognize {actual} as a tool version number.");
+}
+
 namespace vcpkg
 {
-    struct ToolData
+    bool operator==(const ToolVersion& lhs, const ToolVersion& rhs) { return lhs.version == rhs.version; }
+    bool operator<(const ToolVersion& lhs, const ToolVersion& rhs) { return lhs.version < rhs.version; }
+
+    ToolVersion ToolVersion::from_values(std::initializer_list<uint64_t> values)
     {
-        std::array<int, 3> version;
-        Path exe_path;
-        std::string url;
-        Path download_path;
-        bool is_archive;
-        Path tool_dir_path;
-        std::string sha512;
-    };
+        ToolVersion result;
+        result.version = values;
+        auto first = values.begin();
+        const auto last = values.end();
+        if (first != last)
+        {
+            fmt::format_to(std::back_inserter(result.original_text), FMT_STRING("{}"), *first);
+            while (++first != last)
+            {
+                fmt::format_to(std::back_inserter(result.original_text), FMT_STRING(".{}"), *first);
+            }
+        }
+
+        return result;
+    }
 
-    // /\d+\.\d+(\.\d+)?/
-    Optional<std::array<int, 3>> parse_tool_version_string(StringView string_version)
+    ExpectedL<ToolVersion> ToolVersion::try_parse_git(StringView git_version)
+    {
+        /* Sample output:
+git version 2.17.1.windows.2
+
+Expected result: 2.17.1.2
+*/
+        constexpr StringLiteral prefix = "git version ";
+        auto first = git_version.begin();
+        const auto last = git_version.end();
+
+        first = std::search(first, last, prefix.begin(), prefix.end());
+        if (first != last)
+        {
+            first += prefix.size();
+        }
+
+        std::string accumulator;
+        for (; first != last; ++first)
```

Comment:
It might be good to break on whitespace here, just to avoid concerns about globbing a bunch of bonus dots and numbers from multiple lines of output.